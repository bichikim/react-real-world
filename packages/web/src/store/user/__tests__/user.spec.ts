import {ApolloError, ApolloProvider} from '@apollo/client'
import flush from 'flush-promises'
import {client} from 'src/__tests__/apollo-clinet-for-testing'
import mockResult from 'src/store/products/__tests__/porducts.json'
import {act, renderHook} from '@testing-library/react-hooks'
import {createElement as h} from 'react'
import {createUser, user, useUser} from '..'
import couponsMock from './coupons.json'

const setup = () => {
  return createUser(client)
}

const setupUse = () => {
  const wrapper = renderHook(() => (useUser()), {
    wrapper: (props) => (h(ApolloProvider, {client}, props.children)),
  })

  return {
    ...wrapper,
  }
}

describe('user', function test() {
  describe('createUser', function test() {

    beforeEach(() => {
      fetchMock.resetMocks()
      user.state.coupons.clear()
    })

    it('should have all members', function test() {
      const {state, getCoupon, requestGetCoupons, updateState, addCoupons} = setup()
      expect(state).toBeObject()
      expect(getCoupon).toBeFunction()
      expect(requestGetCoupons).toBeFunction()
      expect(updateState).toBeFunction()
      expect(addCoupons).toBeFunction()
    })

    describe('requestGetCoupons', function test() {
      it('should request get coupons', async function test() {
        const {state, requestGetCoupons} = setup()

        expect(state.coupons.size).toBe(0)
        expect(state.state).toBe('idle')

        fetchMock.mockResponseOnce(JSON.stringify(couponsMock))

        requestGetCoupons({})

        expect(state.state).toBe('loading')

        await flush()

        expect(state.state).toBe('idle')

        expect([...state.coupons.entries()]).toMatchSnapshot()

        fetchMock.mockRejectOnce(new ApolloError({}))

        requestGetCoupons({})

        expect(state.state).toBe('loading')

        await flush()

        expect(state.state).toBe('error')
      })
    })

    describe('getCoupon', function test() {
      it('should get on coupon', async function test() {
        const {state, getCoupon, requestGetCoupons} = setup()

        expect(state.coupons.size).toBe(0)

        fetchMock.mockResponseOnce(JSON.stringify(mockResult))

        requestGetCoupons({})

        await flush()

        expect(getCoupon('1')).toMatchSnapshot()
      })
    })
  })
  describe('useUser', function test() {
    it('should work like createUser', async function test() {
      const {result} = setupUse()

      fetchMock.mockResponseOnce(JSON.stringify(couponsMock))

      act(() => {
        result.current.requestGetCoupons({})
      })

      await flush()

      expect(result.current.state.coupons.size).toEqual(2)
    })
  })
})
