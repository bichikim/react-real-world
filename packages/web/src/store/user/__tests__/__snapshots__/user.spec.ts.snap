// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user createUser getCoupon should get on coupon 1`] = `undefined`;

exports[`user createUser requestGetCoupons should request get coupons 1`] = `
Array [
  Array [
    "1",
    Object {
      "discountRate": 10,
      "id": "1",
      "title": "10% 할인 쿠폰",
      "type": "rate",
    },
  ],
  Array [
    "2",
    Object {
      "discountAmount": 10000,
      "id": "2",
      "title": "10,000원 할인 쿠폰",
      "type": "amount",
    },
  ],
]
`;

exports[`user useUser getCoupon should get on coupon 1`] = `undefined`;

exports[`user useUser requestGetCoupons should request get coupons 1`] = `
Array [
  Array [
    "1",
    Object {
      "discountRate": 10,
      "id": "1",
      "title": "10% 할인 쿠폰",
      "type": "rate",
    },
  ],
  Array [
    "2",
    Object {
      "discountAmount": 10000,
      "id": "2",
      "title": "10,000원 할인 쿠폰",
      "type": "amount",
    },
  ],
]
`;
